components/form-input/src/FormInput.js

```diff
@@ -0,0 +1,202 @@
+/** @jsx jsx */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import { jsx, useTheme } from '@westpac/core';
+
+import svgToTinyDataURI from 'mini-svg-data-uri';
+
+// ==============================
+// Utils
+// ==============================
+
+// ==============================
+// Component
+// ==============================
+
+export const FormInput = ({
+	size,
+	width,
+	invalid,
+	tag: Tag,
+	spacing,
+	inline,
+	children,
+	...props
+}) => {
+	const { colors, formInput, typography } = useTheme();
+
+	// Common styling
+	const styleCommon = {
+		display: inline ? 'inline-block' : 'block',
+		width: inline ? 'auto' : '100%',
+		appearance: 'none',
+		lineHeight: formInput.lineHeight,
+		fontWeight: formInput.fontWeight,
+		color: formInput.color,
+		backgroundColor: formInput.backgroundColor,
+		border: `${formInput.borderWidth}px solid`,
+		borderColor:
+			invalid || props.ariaInvalid
+				? formInput.appearance.invalid.borderColor
+				: formInput.appearance.default.borderColor,
+		borderRadius: formInput.borderRadius,
+		transition: 'border 0.2s ease',
+		verticalAlign: Tag === 'textarea' ? 'top' : inline ? 'middle' : null,
+
+		'&::placeholder': {
+			opacity: 1, // Override Firefox's unusual default opacity
+			...formInput.placeholder,
+		},
+
+		// Focus styling (for all, not just keyboard users)
+		':focus': {
+			...typography.link.focus,
+		},
+
+		// Focus styling (for all, not just keyboard users)
+		':focus': {
+			...typography.link.focus,
+		},
+
+		// Disabled and read-only inputs
+		':disabled, &[readonly]': {
+			cursor: 'not-allowed',
+			opacity: 1, // iOS fix for unreadable disabled content
+			...formInput.disabled,
+		},
+
+		// Disable number input spinners/steppers
+		'&::-webkit-outer-spin-button, &::-webkit-inner-spin-button': {
+			margin: 0,
+			appearance: 'none',
+		},
+	};
+
+	// Input size styling
+	const styleSize = {
+		padding: formInput.size[size].padding.join(' '),
+		fontSize: formInput.size[size].fontSize,
+		height: `calc(${formInput.lineHeight}em + ${(p => `${p[0]} + ${p[2] || p[0]}`)(
+			formInput.size[size].padding
+		)} + ${formInput.borderWidth * 2}px)`,
+	};
+
+	// Input fixed width styling
+	const styleFixedWidth = () => {
+		const factor = formInput.fontXFactor; //'W' compared to 'x' character (relative to font)
+		const paddings = parseInt(formInput.size[size].padding[1], 10) * 2;
+		const borders = parseInt(formInput.borderWidth, 10) * 2;
+		const extras = paddings + borders;
+
+		return width
+			? {
+					maxWidth: `calc(${extras}px + ${(width * factor).toFixed(1)}ex)`,
+			  }
+			: null;
+	};
+
+	// Select element styling
+	const styleSelect = () => {
+		const caretWidth = 14;
+		const caretSVG = `<svg xmlns="http://www.w3.org/2000/svg" width="${caretWidth}" height="8" viewBox="0 0 14 8"><path fill="${
+			colors.muted
+		}" fill-rule="evenodd" d="M0 0l7 8 7-8z"/></svg>`;
+		const caretGap = formInput.select.caretGap;
```

Comment:
unnecessary variable just use in the calculation. To help with the long variable names destructure again. i.e.
```
const { colors, formInput, typography } = useTheme();
const { select, textarea } = formInput;
```